### ISSUE TITLE
NullPointerException was caught during scan with compile_commands.json file

### ISSUE THREAD

**Issue by axqd** [blackducksoftware/hub-detect#342]
## Expected behavior
* No exception

## Actual behavior
* Caught exception as below:
```
2018-10-17 11:08:46 ERROR [main] --- Exception:
java.lang.NullPointerException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_181]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_181]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_181]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_181]
        at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598) ~[na:1.8.0_181]
        at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677) ~[na:1.8.0_181]
        at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735) ~[na:1.8.0_181]
        at java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:714) ~[na:1.8.0_181]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233) ~[na:1.8.0_181]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[na:1.8.0_181]
        at com.blackducksoftware.integration.hub.detect.bomtool.clang.ClangExtractor.extract(ClangExtractor.java:92) ~[classes!/:na]
        at com.blackducksoftware.integration.hub.detect.bomtool.clang.ClangBomTool.extract(ClangBomTool.java:84) ~[classes!/:na]
        at com.blackducksoftware.integration.hub.detect.workflow.extraction.ExtractionManager.extract(ExtractionManager.java:98) [classes!/:na]
        at com.blackducksoftware.integration.hub.detect.workflow.extraction.ExtractionManager.extract(ExtractionManager.java:67) [classes!/:na]
        at com.blackducksoftware.integration.hub.detect.workflow.extraction.ExtractionManager.performExtractions(ExtractionManager.java:121) [classes!/:na]
        at com.blackducksoftware.integration.hub.detect.workflow.DetectProjectManager.createDetectProject(DetectProjectManager.java:104) [classes!/:na]
        at com.blackducksoftware.integration.hub.detect.Application.runDetect(Application.java:222) [classes!/:na]
        at com.blackducksoftware.integration.hub.detect.Application.run(Application.java:136) [classes!/:na]
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:788) [spring-boot-2.0.3.RELEASE.jar!/:2.0.3.RELEASE]
        at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:778) [spring-boot-2.0.3.RELEASE.jar!/:2.0.3.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:335) [spring-boot-2.0.3.RELEASE.jar!/:2.0.3.RELEASE]
        at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:137) [spring-boot-2.0.3.RELEASE.jar!/:2.0.3.RELEASE]
        at com.blackducksoftware.integration.hub.detect.Application.main(Application.java:125) [classes!/:na]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_181]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_181]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_181]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_181]
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) [hub-detect-4.3.2.jar:na]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) [hub-detect-4.3.2.jar:na]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:50) [hub-detect-4.3.2.jar:na]
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51) [hub-detect-4.3.2.jar:na]
Caused by: java.lang.NullPointerException: null
        at com.blackducksoftware.integration.hub.detect.bomtool.clang.DependenciesListFileManager.getCompilerCommand(DependenciesListFileManager.java:120) ~[classes!/:na]
        at com.blackducksoftware.integration.hub.detect.bomtool.clang.DependenciesListFileManager.generate(DependenciesListFileManager.java:71) ~[classes!/:na]
        at com.blackducksoftware.integration.hub.detect.bomtool.clang.DependenciesListFileManager.generateDependencyFilePaths(DependenciesListFileManager.java:60) ~[classes!/:na]
        at com.blackducksoftware.integration.hub.detect.bomtool.clang.ClangExtractor.lambda$compileCommandToDependencyFilePathsConverter$0(ClangExtractor.java:112) ~[classes!/:na]
        at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267) ~[na:1.8.0_181]
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948) ~[na:1.8.0_181]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[na:1.8.0_181]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[na:1.8.0_181]
        at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:747) ~[na:1.8.0_181]
        at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:721) ~[na:1.8.0_181]
        at java.util.stream.AbstractTask.compute(AbstractTask.java:316) ~[na:1.8.0_181]
        at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731) ~[na:1.8.0_181]
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[na:1.8.0_181]
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[na:1.8.0_181]
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) ~[na:1.8.0_181]
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[na:1.8.0_181]
```
## Steps to Reproduce
* Generate the compile_commands.json file with either 'compiledb' or 'Bear' for a makefile-based C/C++ project
* Perform the hub-detect scan

## Version
**Project Version:**
* 4.3.2

**Language Version:**
* openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)

**OS:**
* Centos 7

**Comment by stevebillings** [comment]
Thanks axqd, I have been able to reproduce this here using compiledb. I hadn't realized that the command field was optional. I will add support for the case where command is absent, and arguments is present. How urgently do you need a fix?

**Comment by axqd** [comment]
I tried again with the version 4.4.2 and it seems the exception is gone. 
Can you help to confirm this issue has been fixed by the latest version?

**Comment by akamen** [comment]
The issue is fixed in 4.4.2 per release notes.
